File: expo-app/lib/logic/tripSummaryCalculator.ts
/**
 * TripSummaryCalculator - Calculates comprehensive trip statistics
 * 
 * This is a direct port of TripSummaryCalculator.swift with identical behavior.
 * All amounts are in cents (integers) to avoid floating-point issues.
 */

import { v4 as uuidv4 } from 'uuid';
import {
  Trip,
  TripExpense,
  TripParticipant,
  Transaction,
  Category,
  TripSummary,
  CategorySpending,
} from './types';

/**
 * Calculate comprehensive summary for a trip
 */
export function calculate(trip: Trip, expenses: TripExpense[]): TripSummary {
  const transactions = expenses
    .map(e => e.transaction)
    .filter((t): t is Transaction => t !== undefined);
  const participants = trip.participants ?? [];

  // Total spent
  const totalSpentCents = transactions.reduce((sum, tx) => sum + tx.amountCents, 0);

  // Days calculation
  const { daysCount, dailyAverageCents } = calculateDailyStats(
    trip,
    transactions,
    totalSpentCents
  );

  // Budget calculations
  let budgetRemainingCents: number | undefined;
  let budgetPercentUsed: number | undefined;

  if (trip.budgetCents !== undefined && trip.budgetCents > 0) {
    budgetRemainingCents = trip.budgetCents - totalSpentCents;
    budgetPercentUsed = (totalSpentCents / trip.budgetCents) * 100;
  }

  return {
    totalSpentCents,
    dailyAverageCents,
    daysCount,
    budgetRemainingCents,
    budgetPercentUsed,
    expenseCount: transactions.length,
    participantCount: participants.length,
  };
}

/**
 * Calculate spending breakdown by category
 */
export function categorySpending(
  expenses: TripExpense[],
  categoryLookup: Record<string, Category>
): CategorySpending[] {
  const transactions = expenses
    .map(e => e.transaction)
    .filter((t): t is Transaction => t !== undefined);

  const totalCents = transactions.reduce((sum, tx) => sum + tx.amountCents, 0);

  if (totalCents === 0) {
    return [];
  }

  // Group by category
  const categoryTotals: Record<string, { amountCents: number; count: number }> = {};

  for (const tx of transactions) {
    const catId = tx.categoryId ?? 'uncategorized';
    if (!categoryTotals[catId]) {
      categoryTotals[catId] = { amountCents: 0, count: 0 };
    }
    categoryTotals[catId].amountCents += tx.amountCents;
    categoryTotals[catId].count += 1;
  }

  return Object.entries(categoryTotals)
    .map(([catId, data]) => {
      const category = catId !== 'uncategorized' ? categoryLookup[catId] : undefined;
      const percentage = (data.amountCents / totalCents) * 100;

      return {
        id: category?.id ?? uuidv4(),
        category,
        amountCents: data.amountCents,
        percentage,
        transactionCount: data.count,
      };
    })
    .sort((a, b) => b.amountCents - a.amountCents);
}

/**
 * Get display name for a category spending entry
 */
export function getCategoryDisplayName(spending: CategorySpending): string {
  return spending.category?.name ?? 'Uncategorized';
}

/**
 * Get emoji for a category spending entry
 */
export function getCategoryEmoji(spending: CategorySpending): string {
  return spending.category?.emoji ?? '📝';
}

/**
 * Calculate spending by participant (what each person spent for the group)
 */
export function participantSpending(
  expenses: TripExpense[],
  participantLookup: Record<string, TripParticipant>
): Array<{ participant: TripParticipant; amountCents: number }> {
  const spending: Record<string, number> = {};

  for (const expense of expenses) {
    const participantId = expense.paidByParticipantId;
    const transaction = expense.transaction;

    if (!participantId || !transaction) continue;

    spending[participantId] = (spending[participantId] ?? 0) + transaction.amountCents;
  }

  return Object.entries(spending)
    .map(([participantId, amountCents]) => ({
      participant: participantLookup[participantId],
      amountCents,
    }))
    .filter(entry => entry.participant !== undefined)
    .sort((a, b) => b.amountCents - a.amountCents);
}

/**
 * Calculate daily spending breakdown
 */
export function dailySpending(
  expenses: TripExpense[]
): Array<{ dateMs: number; amountCents: number }> {
  const transactions = expenses
    .map(e => e.transaction)
    .filter((t): t is Transaction => t !== undefined);

  const dailyTotals: Record<string, number> = {};

  for (const tx of transactions) {
    // Get date at midnight (start of day)
    const date = new Date(tx.date);
    const dayKey = new Date(date.getFullYear(), date.getMonth(), date.getDate()).getTime().toString();

    dailyTotals[dayKey] = (dailyTotals[dayKey] ?? 0) + tx.amountCents;
  }

  return Object.entries(dailyTotals)
    .map(([dateMs, amountCents]) => ({
      dateMs: parseInt(dateMs, 10),
      amountCents,
    }))
    .sort((a, b) => a.dateMs - b.dateMs);
}

/**
 * Find the highest expense in a trip
 */
export function highestExpense(
  expenses: TripExpense[]
): { transaction: Transaction; amountCents: number } | undefined {
  const transactions = expenses
    .map(e => e.transaction)
    .filter((t): t is Transaction => t !== undefined);

  if (transactions.length === 0) {
    return undefined;
  }

  const highest = transactions.reduce((max, tx) =>
    tx.amountCents > max.amountCents ? tx : max
  );

  return {
    transaction: highest,
    amountCents: highest.amountCents,
  };
}

/**
 * Find the most expensive day
 */
export function mostExpensiveDay(
  expenses: TripExpense[]
): { dateMs: number; amountCents: number } | undefined {
  const daily = dailySpending(expenses);

  if (daily.length === 0) {
    return undefined;
  }

  return daily.reduce((max, day) =>
    day.amountCents > max.amountCents ? day : max
  );
}

// Private helper

function calculateDailyStats(
  trip: Trip,
  transactions: Transaction[],
  totalCents: number
): { daysCount: number; dailyAverageCents: number } {
  // Determine date range
  const transactionDates = transactions.map(tx => tx.date);
  const minDate = transactionDates.length > 0
    ? Math.min(...transactionDates)
    : (trip.startDate ?? Date.now());
  const maxDate = transactionDates.length > 0
    ? Math.max(...transactionDates)
    : (trip.endDate ?? Date.now());

  // If trip has explicit dates, use those
  const startDate = trip.startDate ?? minDate;
  const endDate = trip.endDate ?? maxDate;

  // Calculate days (inclusive)
  const msPerDay = 24 * 60 * 60 * 1000;
  const days = Math.max(1, Math.floor((endDate - startDate) / msPerDay) + 1);

  const dailyAvg = Math.round(totalCents / days);

  return { daysCount: days, dailyAverageCents: dailyAvg };
}

// Export as namespace for compatibility
export const TripSummaryCalculator = {
  calculate,
  categorySpending,
  getCategoryDisplayName,
  getCategoryEmoji,
  participantSpending,
  dailySpending,
  highestExpense,
  mostExpensiveDay,
};

File: legacySwiftApp/BudgetThing/TripSummaryCalculator.swift
//
//  TripSummaryCalculator.swift
//  BudgetThing
//

import Foundation

// MARK: - Data Structures

struct TripSummary {
    let totalSpent: Decimal
    let spentByCategory: [(category: Category?, amount: Decimal)]
    let dailyAverage: Decimal
    let daysCount: Int
    let budgetRemaining: Decimal?
    let budgetPercentUsed: Double?
    let expenseCount: Int
    let participantCount: Int
    let highestExpense: (transaction: Transaction, amount: Decimal)?
    let mostExpensiveDay: (date: Date, amount: Decimal)?
}

struct CategorySpending: Identifiable {
    let id: UUID
    let category: Category?
    let amount: Decimal
    let percentage: Double
    let transactionCount: Int

    var displayName: String {
        category?.name ?? "Uncategorized"
    }

    var emoji: String {
        category?.emoji ?? "📝"
    }
}

// MARK: - Summary Calculator

enum TripSummaryCalculator {

    /// Calculate comprehensive summary for a trip
    static func calculate(trip: Trip) -> TripSummary {
        let expenses = trip.expenses ?? []
        let transactions = expenses.compactMap { $0.transaction }
        let participants = trip.participants ?? []

        // Total spent
        let total = transactions.reduce(Decimal(0)) { $0 + $1.amount }

        // Group by category
        let byCategory = calculateCategorySpending(transactions: transactions)

        // Days calculation
        let (daysCount, dailyAvg) = calculateDailyStats(
            trip: trip,
            transactions: transactions,
            total: total
        )

        // Budget calculations
        let budgetRemaining = trip.budget.map { $0 - total }
        let budgetPercent: Double? = trip.budget.flatMap { budget in
            guard budget > 0 else { return nil }
            return Double(truncating: (total / budget * 100) as NSDecimalNumber)
        }

        // Highest expense
        let highestExpense = transactions.max(by: { $0.amount < $1.amount })
            .map { ($0, $0.amount) }

        // Most expensive day
        let mostExpensiveDay = calculateMostExpensiveDay(transactions: transactions)

        return TripSummary(
            totalSpent: total,
            spentByCategory: byCategory,
            dailyAverage: dailyAvg,
            daysCount: daysCount,
            budgetRemaining: budgetRemaining,
            budgetPercentUsed: budgetPercent,
            expenseCount: transactions.count,
            participantCount: participants.count,
            highestExpense: highestExpense,
            mostExpensiveDay: mostExpensiveDay
        )
    }

    /// Calculate spending breakdown by category
    static func categorySpending(trip: Trip) -> [CategorySpending] {
        let expenses = trip.expenses ?? []
        let transactions = expenses.compactMap { $0.transaction }
        let total = transactions.reduce(Decimal(0)) { $0 + $1.amount }

        guard total > 0 else { return [] }

        // Group by category
        var categoryTotals: [UUID?: (category: Category?, amount: Decimal, count: Int)] = [:]

        for tx in transactions {
            let catId = tx.category?.id
            let existing = categoryTotals[catId]
            categoryTotals[catId] = (
                category: tx.category ?? existing?.category,
                amount: (existing?.amount ?? 0) + tx.amount,
                count: (existing?.count ?? 0) + 1
            )
        }

        return categoryTotals.map { _, value in
            let percentage = Double(truncating: (value.amount / total * 100) as NSDecimalNumber)
            return CategorySpending(
                id: value.category?.id ?? UUID(),
                category: value.category,
                amount: value.amount,
                percentage: percentage,
                transactionCount: value.count
            )
        }.sorted { $0.amount > $1.amount }
    }

    /// Calculate spending by participant (what each person spent for the group)
    static func participantSpending(trip: Trip) -> [(participant: TripParticipant, amount: Decimal)] {
        let expenses = trip.expenses ?? []
        var spending: [UUID: (participant: TripParticipant, amount: Decimal)] = [:]

        for expense in expenses {
            guard let participant = expense.paidByParticipant,
                  let transaction = expense.transaction else { continue }

            let existing = spending[participant.id]
            spending[participant.id] = (
                participant: participant,
                amount: (existing?.amount ?? 0) + transaction.amount
            )
        }

        return spending.values.sorted { $0.amount > $1.amount }
    }

    /// Calculate daily spending breakdown
    static func dailySpending(trip: Trip) -> [(date: Date, amount: Decimal)] {
        let expenses = trip.expenses ?? []
        let transactions = expenses.compactMap { $0.transaction }

        let calendar = Calendar.current
        var dailyTotals: [DateComponents: Decimal] = [:]

        for tx in transactions {
            let components = calendar.dateComponents([.year, .month, .day], from: tx.date)
            dailyTotals[components, default: 0] += tx.amount
        }

        return dailyTotals.compactMap { components, amount in
            guard let date = calendar.date(from: components) else { return nil }
            return (date, amount)
        }.sorted { $0.date < $1.date }
    }

    // MARK: - Private Helpers

    private static func calculateCategorySpending(
        transactions: [Transaction]
    ) -> [(category: Category?, amount: Decimal)] {
        var categoryTotals: [UUID?: Decimal] = [:]
        var categoryLookup: [UUID?: Category?] = [:]

        for tx in transactions {
            let catId = tx.category?.id
            categoryTotals[catId, default: 0] += tx.amount
            if categoryLookup[catId] == nil {
                categoryLookup[catId] = tx.category
            }
        }

        return categoryTotals.map { catId, amount in
            (categoryLookup[catId] ?? nil, amount)
        }.sorted { $0.amount > $1.amount }
    }

    private static func calculateDailyStats(
        trip: Trip,
        transactions: [Transaction],
        total: Decimal
    ) -> (daysCount: Int, dailyAverage: Decimal) {
        let calendar = Calendar.current

        // Determine date range
        let transactionDates = transactions.map { $0.date }
        let minDate = transactionDates.min() ?? trip.startDate ?? Date()
        let maxDate = transactionDates.max() ?? trip.endDate ?? Date()

        // If trip has explicit dates, use those
        let startDate = trip.startDate ?? minDate
        let endDate = trip.endDate ?? maxDate

        let days = max(1, (calendar.dateComponents([.day], from: startDate, to: endDate).day ?? 0) + 1)
        let dailyAvg = total / Decimal(days)

        return (days, dailyAvg)
    }

    private static func calculateMostExpensiveDay(
        transactions: [Transaction]
    ) -> (date: Date, amount: Decimal)? {
        let calendar = Calendar.current
        var dailyTotals: [DateComponents: Decimal] = [:]

        for tx in transactions {
            let components = calendar.dateComponents([.year, .month, .day], from: tx.date)
            dailyTotals[components, default: 0] += tx.amount
        }

        guard let maxEntry = dailyTotals.max(by: { $0.value < $1.value }),
              let date = calendar.date(from: maxEntry.key) else {
            return nil
        }

        return (date, maxEntry.value)
    }
}
